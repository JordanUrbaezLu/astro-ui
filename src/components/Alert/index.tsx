/* istanbul ignore file */
export * from "./Alert";
