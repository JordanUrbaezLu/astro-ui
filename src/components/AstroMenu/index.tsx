export * from "./AstroMenu";
export * from "./AstroMenuItem";
