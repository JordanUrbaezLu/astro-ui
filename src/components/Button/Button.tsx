import * as React from "react";
import styles from "./Button.module.scss";
import classNames from "classnames";
import { Typography } from "../Typography/Typography";
import { Ripple } from "../Ripple";

export type ButtonVariant = "primary" | "secondary";
export type ButtonSize = "small" | "medium" | "large";
export type ButtonType = "button" | "submit" | "reset";

export interface ButtonProps
  extends React.ComponentPropsWithoutRef<"button"> {
  /**
   * The content for the Button
   */
  children: string;
  /**
   * If the Button is disabled
   *
   * @default false
   */
  disabled?: boolean;
  /**
   * The leading icon for the Button
   */
  leadingIcon?: React.ReactNode;
  /**
   * The callback fired when the Button is clicked
   */
  onClick?: () => void;
  /**
   * The size of the Button
   *
   * @default "medium"
   */
  size?: ButtonSize;
  /**
   * The trailing icon for the Button
   */
  trailingIcon?: React.ReactNode;
  /**
   * The type of the Button
   */
  type?: ButtonType;
  /**
   * The variant of the Button
   *
   * @default "primary"
   */
  variant?: ButtonVariant;
}

/**
 * @public
 *
 * @description
 *
 * The Button component is used to allows user to take an action.
 */
export const Button = React.forwardRef<
  HTMLButtonElement,
  ButtonProps
>((props, ref) => {
  const {
    className,
    children,
    disabled = false,
    leadingIcon,
    onClick,
    size = "medium",
    trailingIcon,
    type,
    variant = "primary",
    ...rest
  } = props;

  return (
    <button
      className={classNames(
        className,
        styles.button,
        disabled && styles.disabled,
        size === "small"
          ? styles.small
          : size === "large"
          ? styles.large
          : styles.medium,
        variant === "secondary" ? styles.secondary : styles.primary
      )}
      disabled={disabled}
      onClick={onClick}
      ref={ref}
      type={
        type === "reset"
          ? "reset"
          : type === "submit"
          ? "submit"
          : "button"
      }
      {...rest}
    >
      {!disabled && <Ripple position="cursor" />}
      <Typography
        className={styles.typography}
        type={
          size === "large" ? "h1" : size === "medium" ? "h2" : "h3"
        }
        variant={variant}
      >
        {leadingIcon && (
          <span className={styles.leadingIcon}>{leadingIcon}</span>
        )}
        {children}
        {trailingIcon && (
          <span className={styles.trailingIcon}>{trailingIcon}</span>
        )}
      </Typography>
    </button>
  );
});

Button.displayName = "Button";
