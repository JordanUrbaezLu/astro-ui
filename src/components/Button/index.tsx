/* istanbul ignore file */
export * from "./Button";
