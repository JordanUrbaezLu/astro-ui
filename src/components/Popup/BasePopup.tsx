import classNames from "classnames";
import * as React from "react";
import { Typography } from "../Typography/Typography";

import styles from "./BasePopup.module.scss";

export type BasePopupPosition = "bottom" | "top" | "left" | "right";

export interface BasePopupProps
  extends React.ComponentPropsWithoutRef<"div"> {
  /**
   * If the Popup has an indicator
   */
  hasIndicator: boolean;
  /**
   * The position of the Popup
   */
  position: BasePopupPosition;
  /**
   * The text for the Popup
   */
  text: string;
}

export const BasePopup = React.forwardRef<
  HTMLDivElement,
  BasePopupProps
>((props, ref) => {
  const { className, hasIndicator, position, text, ...rest } = props;

  return (
    <div
      className={classNames(
        hasIndicator && styles.indicator,
        position === "bottom" && styles.bottom,
        position === "top" && styles.top,
        position === "left" && styles.left,
        position === "right" && styles.right,
        styles.basePopup,
        className
      )}
      ref={ref}
      role="tooltip"
      {...rest}
    >
      <Typography type="h2" variant="primary">
        {text}
      </Typography>
    </div>
  );
});

BasePopup.displayName = "BasePopup";
