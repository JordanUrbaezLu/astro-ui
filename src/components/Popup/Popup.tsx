import * as React from "react";
import { CSSTransition } from "react-transition-group";
import styles from "./Popup.module.scss";
import classNames from "classnames";
import {
  useOnFocus,
  useOnFocusOut,
  useOnHover,
  useOnHoverOut,
} from "../../hooks";
import {
  BasePopup,
  BasePopupPosition as PopupPosition,
} from "./BasePopup";
import { PopupLayoutContainer } from "./PopupLayoutContainer";

export interface PopupProps
  extends React.ComponentPropsWithoutRef<"div"> {
  /**
   * The trigger for the Popup
   */
  children: React.ReactElement;
  /**
   * If the Popup has an indicator
   *
   * @default false
   */
  hasIndicator?: boolean;
  /**
   * If the Popup is open
   *
   * @default false
   */
  isOpen?: boolean;
  /**
   * The callback fired when the Popup closes
   */
  onClose: (itemValue?: string) => void;
  /**
   * The callback fired when the Popup opens
   */
  onOpen: () => void;
  /**
   * The position of the Popup
   *
   * @default "bottom"
   */
  position?: PopupPosition;
  /**
   * The text for the Popup
   */
  text: string;
}

/**
 * @public
 *
 * @description
 *
 * The Popup component is a display that shows when hovering
 * over an element.
 */
export const Popup: React.FC<PopupProps> = (props) => {
  const {
    className,
    children,
    hasIndicator = false,
    isOpen = false,
    onClose,
    onOpen,
    position = "bottom",
    text,
    ...rest
  } = props;

  const popupRef = React.useRef<HTMLDivElement>(null);

  const triggerRef = React.useRef<HTMLDivElement>(null);

  // Show Popup on focus or hover
  useOnFocus(triggerRef, onOpen);
  useOnHover(triggerRef, onOpen);

  // Hide Popup on focus out or hover out
  useOnFocusOut(triggerRef, onClose);
  useOnHoverOut(triggerRef, onClose);

  return (
    <div
      className={classNames(
        position === "bottom" && styles.bottom,
        position === "top" && styles.top,
        position === "left" && styles.left,
        position === "right" && styles.right,
        styles.container
      )}
    >
      {React.cloneElement(React.Children.only(children), {
        ref: triggerRef,
        role: "button",
        tabIndex: 0,
      })}

      <CSSTransition
        classNames={{
          enter: styles.enter,
          enterActive: styles.enterActive,
          exit: styles.exit,
          exitActive: styles.exitActive,
        }}
        in={isOpen}
        mountOnEnter
        nodeRef={popupRef}
        timeout={200}
        unmountOnExit
      >
        <PopupLayoutContainer
          hasIndicator={hasIndicator}
          position={position}
          triggerRef={triggerRef}
        >
          <BasePopup
            className={className}
            hasIndicator={hasIndicator}
            position={position}
            ref={popupRef}
            text={text}
            {...rest}
          />
        </PopupLayoutContainer>
      </CSSTransition>
    </div>
  );
};

Popup.displayName = "Popup";
