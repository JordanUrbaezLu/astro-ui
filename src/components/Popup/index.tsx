/* istanbul ignore file */
export * from "./Popup";
