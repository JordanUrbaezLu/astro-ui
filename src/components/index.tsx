export * from "./Alert";
export * from "./Button";
export * from "./Checkbox";
export * from "./DatePicker";
export * from "./Link";
export * from "./Menu";
export * from "./Tooltip";
export * from "./Select";
export * from "./Switch";
export * from "./TextField";
