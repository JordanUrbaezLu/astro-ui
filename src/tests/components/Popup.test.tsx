import { render, screen } from "@testing-library/react";
import { Popup } from "../../components/Popup";
import { axe } from "jest-axe";
import { Button } from "../../components/Button";
import { BasePopupPosition } from "../../components/Popup/BasePopup";

describe("Popup", () => {
  test("Should render component correctly.", () => {
    const { container } = render(
      <Popup onClose={jest.fn()} onOpen={jest.fn()} text="Popup">
        <Button>Trigger</Button>
      </Popup>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(
      <Popup onClose={jest.fn()} onOpen={jest.fn()} text="Popup">
        <Button>Trigger</Button>
      </Popup>
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe("Interaction", () => {
  test("Should trigger onOpen and onClose correctly.", () => {
    const onOpen = jest.fn();
    const onClose = jest.fn();

    render(
      <>
        <button>Button</button>
        <Popup onClose={onClose} onOpen={onOpen} text="Popup">
          <Button>Trigger</Button>
        </Popup>
      </>
    );

    screen.getAllByRole("button")[1].focus();

    expect(onOpen).toHaveBeenCalledTimes(1);

    screen.getAllByRole("button")[0].focus();

    expect(onClose).toHaveBeenCalledTimes(1);
  });
});

describe("Props", () => {
  test.each<BasePopupPosition>(["bottom", "top", "left", "right"])(
    "Should render position correctly.",
    (position) => {
      render(
        <Popup
          isOpen
          onClose={jest.fn()}
          onOpen={jest.fn()}
          position={position}
          text="Popup"
        >
          <Button>Trigger</Button>
        </Popup>
      );
      expect(screen.getByRole("tooltip")).toHaveClass(position);
    }
  );

  test("Should render indicator correctly.", () => {
    render(
      <Popup
        hasIndicator
        isOpen
        onClose={jest.fn()}
        onOpen={jest.fn()}
        text="Popup"
      >
        <Button>Trigger</Button>
      </Popup>
    );
    expect(screen.getByRole("tooltip")).toHaveClass("indicator");
  });
});
