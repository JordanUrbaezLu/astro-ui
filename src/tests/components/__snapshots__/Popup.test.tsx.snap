// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup Should render component correctly. 1`] = `
<div
  class="bottom container"
>
  <button
    class="button medium primary"
    role="button"
    tabindex="0"
    type="button"
  >
    <div
      class="container"
      role="presentation"
    />
    <div
      class="typography typography typographyh2 primary normal"
    >
      Trigger
    </div>
  </button>
</div>
`;
